plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapHighLevelClient.java

```diff
@@ -0,0 +1,164 @@
+     */
+    public Optional<LdapSearchUserResult> getUser(String user)
+            throws NamingException
+    {
+        return ldapClient.executeLdapQuery(ldapAdminUser, ldapAdminPassword,
+                new LdapQuery.LdapQueryBuilder()
+                        .withSearchBase(userBaseDN)
+                        .withAttributes(userSearchAttributesArray)
+                        .withSearchFilter(userSearchFilter)
+                        .withFilterArguments(user)
+                        .build(),
+                search -> {
+                    if (!search.hasMore()) {
+                        log.debug("LDAP search for user [%s] using filter pattern [%s] found no matches", user, userSearchFilter);
+                        return Optional.empty();
+                    }
+                    SearchResult result = search.next();
+                    ImmutableMap.Builder<String, List<String>> resAttributesBuilder = ImmutableMap.builder();
+                    for (String attributeName : userSearchAttributes) {
+                        Attribute attribute = result.getAttributes().get(attributeName);
+                        if (attribute == null) {
+                            log.warn("User [%s] does not have user search attribute [%s]", user, attributeName);
+                            continue;
+                        }
+                        NamingEnumeration<?> attributeValues = attribute.getAll();
+                        ImmutableList.Builder<String> valuesBuilder = ImmutableList.builder();
+                        while (attributeValues.hasMore()) {
+                            valuesBuilder.add(attributeValues.next().toString());
+                        }
+                        List<String> values = valuesBuilder.build();
+                        if (values.isEmpty()) {
+                            log.debug("User [%s] has attribute [%s] with not values", user, attributeName);
+                            continue;
+                        }
+                        resAttributesBuilder.put(attributeName, values);
+                    }
+                    return Optional.of(new LdapSearchUserResult(result.getNameInNamespace(), resAttributesBuilder.buildOrThrow()));
+                });
+    }
+
+    /**
+     * Perform an LDAP search for groups, fetching only the names, and returning the name of each group.
+     * Filters groups by user membership AND filter expression {@link LdapHighLevelClientConfig#getLdapGroupsSearchFilter()}.
+     * If {@link LdapHighLevelClientConfig#getLdapGroupsNameAttribute()} is missing from group document, fallback on full name.
+     *
+     * @param userDN Fully quantified DN of user, used in filter to bind group attribute given by
+     * {@link LdapHighLevelClientConfig#getLdapGroupsSearchMemberAttribute()}
+     * @return Names of groups that the user is a member of
+     * @throws NamingException Underlying LDAP error
+     */
+    public List<String> getGroups(String userDN)
```

Comment:
Apart from this approach we also have another one - for AD kind of solution i.e if `memberOf` attribute is present then we could search for the user in a single LDAP query. Though it is out of scope for this PR - It would be helpful if we could proper abstraction so that it could also be benefited. WDYT ?

Reply:
We tried using this solution, but ran into two issues:
-  using `memberOf` will fetch ALL groups that a user is a member of. This can be bad for both performance and security reasons;
- `memberOf` attributes _usually_ (depends on the LDAP application) contain the fully qualified name of the group (e.g., `cn=clients,ou=groups,dc=trino,dc=io`)

But I can see there are cases where these are not issues and having a one-query solution would be nice. 

The `LdapGroupsMapping` implementation from the Hadoop libraries allows you to configure a `memberOfAttribute` and will use one query if that is specified - WDYT of doing something similar?